AvalonStudio/AvalonStudio.Extensibility/Utils/FuzzyMatch.cs

```diff
@@ -0,0 +1,448 @@
+            List<Match> matchList = null;
+            var remainder = "";
+            var extraCharacters = specials[startingSpecial] + query.Length - str.Length;
+
+            if (extraCharacters > 0)
+            {
+                remainder = query.Substring(0, extraCharacters);
+                query = query.Substring(extraCharacters);
+            }
+
+            int queryCounter;
+            for (queryCounter = 0; queryCounter < query.Length; queryCounter++)
+            {
+                matchList = generateMatchList(query.Substring(queryCounter), str, specials, startingSpecial);
+
+                if(matchList != null || startingSpecial == 0)
+                {
+                    break;
+                }
+            }
+
+            if(queryCounter == query.Length || matchList == null)
+            {
+                return null;
+            }
+
+            return (remainder + query.Substring(0, queryCounter), matchList);
+        }
+
+
+        static List<Match> _wholeStringSearch(string query, string str, List<int> specials, int lastSegmentSpecialsIndex)
+        {
+            query = query.ToLower();
+            var compareStr = str.ToLower();
+
+            var lastSegmentStart = specials[lastSegmentSpecialsIndex];
+
+            List<Match> matchList = new List<Match>();
+
+            var result = _lastSegmentSearch(query, compareStr, specials, lastSegmentSpecialsIndex, lastSegmentStart);
+
+            if(result.HasValue)
+            {
+                matchList = result.Value.matchList;
+
+                if(result.Value.remainder != "")
+                {
+                    var remainderMatchList = generateMatchList(result.Value.remainder, compareStr.Substring(0, lastSegmentStart), specials.GetRange(0, lastSegmentSpecialsIndex), 0);
+                    if (remainderMatchList != null)
+                    {
+                        // add the new matched ranges to the beginning of the set of ranges we had
+                        matchList.InsertRange(0, remainderMatchList);
+                    }
+                    else
+                    {
+                        // No match
+                        return null;
+                    }
+                } 
+            } else
+            {
+                // no match in last segment, start over searching whole string.
+                matchList = generateMatchList(query, compareStr, specials, 0); // lastSegmentStart as 5th param??
+            }
+
+            return matchList;
+        }
+
+        public class Range
+        {
+            public Range(string text, bool matched, bool includesLastSegment)
+            {
+                this.text = text;
+                this.matched = matched;
+                this.includesLastSegment = includesLastSegment;
+            }
+
+            public string text;
+            public bool matched;
+            public bool includesLastSegment;     
+        };
+
+        const int SPECIAL_POINTS = 35;
```

Comment:
explicityly mark these as private, keep all constants at the top of the file. along side private vars.